### PR TITLE
Faster li, load_const_optimized & add_const_optimized

### PR DESCRIPTION
**Pull request by tsarn**

### CODE REVIEW

File: src/hotspot/cpu/riscv/assembler_riscv.cpp [hunk 1]
```diff
@@ -463,186 +463,173 @@ void Assembler::li(Register d, void* addr) {
...
 
-  short xa, xb, xc, xd; // Four 16-bit chunks of const.
-  long rem = x;         // Remaining part of const.
+  unsigned long uimm = imm;
 
-  xd = rem & 0xFFFF;    // Lowest 16-bit chunk.
-  rem = (rem >> 16) + ((unsigned short)xd >> 15); // Compensation for sign extend.
+  // this loop tries different strategies to load the constant
+  // in the most optimal way
+  for (int i = 0; i < 2; ++i) {
```
Comment (Geckostya) [inline src/hotspot/cpu/riscv/assembler_riscv.cpp]: it should be changed with `if` to check size of constant

File: src/hotspot/cpu/riscv/assembler_riscv.cpp [hunk 2]
```diff
@@ -463,186 +463,173 @@ void Assembler::li(Register d, void* addr) {
...
-      }
-      insrdi_PPC(d, tmp, 32, 0);
-      return retval;
-    }
+  unsigned long sha = 0;
+  while (!(uimm & 1)) {
+    ++sha;
+    uimm >>= 1;
+  }
+  unsigned long low = uimm & 0x3fful;
```
Comment (Geckostya) [inline src/hotspot/cpu/riscv/assembler_riscv.cpp]: can load by 12 bit

**Review by Geckostya — COMMENTED** [review]
It should be tested

**Review by AntonKozlov — COMMENTED** [review]
Ooph, I hope tests passed :)
It's also intersting what is distribution of constants: how many of them is harder than lui+addi?
There is another possibility for "hard" constants (hard definition may vary): load the constant from own codeblob. There is 'Assembler::long_constant` in share/asm/assembler.hpp to generate constants near code stream. Its result apparently is best used with pc-relative addressing (auipc+lui)

**Comment by tsarn** [comment]
> Ooph, I hope tests passed :)

Yeah, they did :)

> It's also intersting what is distribution of constants: how many of them is harder than lui+addi?

Before the VM crashed on unimplemented instructions, all calls to li were either with small immediate value (usually 0) or pc-relative that can be loaded with auipc + addi

**Review by Geckostya — APPROVED** [review]

File: src/hotspot/cpu/riscv/assembler_riscv.cpp [hunk 3]
```diff
@@ -463,186 +463,169 @@ void Assembler::li(Register d, void* addr) {
...
+// load 64-bit immediate value
+void Assembler::li(Register d, long imm) {
+  // tty->print_cr("li %s, 0x%lx at %p", d->name(), (unsigned long)imm, pc());
 
-  for (remBit -= 11; remBit > 0; remBit -= 11) {
-    part = (uimm >> remBit) & 0x7FF;
-    if (part != 0) addi(d, d, part);
-    slli(d, d, remBit >= 11 ? 11 : remBit);
+  if (-0x800 <= imm && imm < 0x800) {
+    addi(d, R0_ZERO, imm);
```
Comment (sreznick) [inline src/hotspot/cpu/riscv/assembler_riscv.cpp:472]: if (imm & 0x7ff) ?

File: src/hotspot/cpu/riscv/assembler_riscv.cpp [hunk 4]
```diff
@@ -463,186 +463,169 @@ void Assembler::li(Register d, void* addr) {
...
+  long off = imm - (long)pc();
 
-// Load a 64 bit constant, optimized, not identifyable.
-// Tmp can be used to increase ILP. Set return_simm16_rest=true to get a
-// 16 bit immediate offset.
-int Assembler::load_const_optimized(Register d, long x, Register tmp, bool return_simm16_rest) {
-  // Avoid accidentally trying to use R0 for indexed addressing.
-  assert_different_registers(d, tmp);
+  if (INT32_MIN <= off && off <= INT32_MAX) {
+    // load using AUIPC
```
Comment (sreznick) [inline src/hotspot/cpu/riscv/assembler_riscv.cpp:479]: if (off & (INT32_MAX - 1))  ?

File: src/hotspot/cpu/riscv/assembler_riscv.cpp [hunk 5]
```diff
@@ -463,186 +463,169 @@ void Assembler::li(Register d, void* addr) {
...
 
-  short xa, xb, xc, xd; // Four 16-bit chunks of const.
-  long rem = x;         // Remaining part of const.
+  unsigned long uimm = imm;
+  unsigned long value = uimm;
 
-  xd = rem & 0xFFFF;    // Lowest 16-bit chunk.
-  rem = (rem >> 16) + ((unsigned short)xd >> 15); // Compensation for sign extend.
+  unsigned long sha = 0; // shift amount for final value
+  if (imm < INT32_MIN || imm > INT32_MAX) {
```
Comment (sreznick) [inline src/hotspot/cpu/riscv/assembler_riscv.cpp:497]: if (imm & ~(INT32_MAX - 1) )

File: src/hotspot/cpu/riscv/assembler_riscv.cpp [hunk 6]
```diff
@@ -463,186 +463,169 @@ void Assembler::li(Register d, void* addr) {
...
+    while (!(value & 1)) {
+      ++sha;
+      value >>= 1;
+    }
+  }
```
Comment (sreznick) [inline src/hotspot/cpu/riscv/assembler_riscv.cpp:503]: to calculate zeros in lowest significant bits there are two alternatives to linear approach implemented here.

For example: value & (value - 1) gives value with single one-valued bit at position of least sugnificant one-bit of value.

And it's offset can be found by binary search.

1.  can be done faster in following way:
1. determine

File: src/hotspot/cpu/riscv/assembler_riscv.cpp [hunk 7]
```diff
@@ -463,186 +463,169 @@ void Assembler::li(Register d, void* addr) {
...
 
-  if (rem == 0) { // opt 1: simm16
-    li_PPC(d, xd);
-    return 0;
+  unsigned long low = value & 0xfff; // low section is 12 lowest bits
+  unsigned long mid = value >> 12; // mid section is 20+ following bits
+  if (low >= 0x800) {
```
Comment (sreznick) [inline src/hotspot/cpu/riscv/assembler_riscv.cpp:506]: 12 and 0xfff are repeated several times.
And 0xfff is (1 << 12) -1.

It makes sense to keep this knowledge in single place. 
For example:

const int lowest_const_bits = 12;
const int mask_lowest_const_bits = (1 << lowest_const_bits) - 1 ;

Comment (sreznick) [reply src/hotspot/cpu/riscv/assembler_riscv.cpp:506]: May be there this such kind of optimized function already in OpenJDK codebase ?

**Review by sreznick — COMMENTED** [review]
I'm done. I found no show stoppers.
All my comments are proposals for some kind of fine-tuning and performance improvements.
They are nor critical. But with such robust framework it is not dangerous to play with some enhancements.

File: src/hotspot/cpu/riscv/assembler_riscv.cpp (hunk 3 above)
Comment (Geckostya) [reply src/hotspot/cpu/riscv/assembler_riscv.cpp:472]: it is not correct working with imm = 0xfffff and so one

File: src/hotspot/cpu/riscv/assembler_riscv.cpp (hunk 4 above)
Comment (Geckostya) [reply src/hotspot/cpu/riscv/assembler_riscv.cpp:479]: the same problem

File: src/hotspot/cpu/riscv/assembler_riscv.cpp (hunk 3 above)
Comment (tsarn) [reply src/hotspot/cpu/riscv/assembler_riscv.cpp:472]: > it is not correct working with imm = 0xfffff and so one

Please elaborate, I tested it and it appears to generate correct code on 0xFFFFF.

**Pull request merged**